src/stack_height/max_height.rs

```diff
@@ -0,0 +1,524 @@
+		}
+	}
+
+	/// Returns current height of the value stack.
+	fn height(&self) -> u32 {
+		self.height
+	}
+
+	/// Returns a reference to a frame by specified depth relative to the top of
+	/// control stack.
+	fn frame(&self, rel_depth: u32) -> Result<&Frame, Error> {
+		let control_stack_height: usize = self.control_stack.len();
+		let last_idx = control_stack_height
+			.checked_sub(1)
+			.ok_or_else(|| Error("control stack is empty".into()))?;
+		let idx = last_idx
+			.checked_sub(rel_depth as usize)
+			.ok_or_else(|| Error("control stack out-of-bounds".into()))?;
+		Ok(&self.control_stack[idx])
+	}
+
+	/// Mark successive instructions as unreachable.
+	///
+	/// This effectively makes stack polymorphic.
+	fn mark_unreachable(&mut self) -> Result<(), Error> {
+		trace!("unreachable");
+		let top_frame = self.control_stack
+			.last_mut()
+			.ok_or_else(|| Error("stack must be non-empty".into()))?;
+		top_frame.is_polymorphic = true;
+		Ok(())
+	}
+
+	/// Push control frame into the control stack.
+	fn push_frame(&mut self, frame: Frame) {
+		trace!("push_frame: {:?}", frame);
+		self.control_stack.push(frame);
+	}
+
+	/// Pop control frame from the control stack.
+	///
+	/// Returns `Err` if the control stack is empty.
+	fn pop_frame(&mut self) -> Result<Frame, Error> {
+		trace!("pop_frame: {:?}", self.control_stack.last());
+		Ok(self.control_stack
+			.pop()
+			.ok_or_else(|| Error("stack must be non-empty".into()))?)
+	}
+
+	/// Truncate the height of value stack to the specified height.
+	fn trunc(&mut self, new_height: u32) {
+		trace!("trunc: {}", new_height);
+		self.height = new_height;
+	}
+
+	/// Push specified number of values into the value stack.
+	///
+	/// Returns `Err` if the height overflow usize value.
+	fn push_values(&mut self, value_count: u32) -> Result<(), Error> {
+		trace!("push: {}", value_count);
+		self.height = self.height
+			.checked_add(value_count)
+			.ok_or_else(|| Error("stack overflow".into()))?;
+		Ok(())
+	}
+
+	/// Pop specified number of values from the value stack.
+	///
+	/// Returns `Err` if the stack happen to be negative value after
+	/// values popped.
+	fn pop_values(&mut self, value_count: u32) -> Result<(), Error> {
+		trace!("pop: {}", value_count);
+		if value_count == 0 {
+			return Ok(());
+		}
+		{
+			let top_frame = self.frame(0)?;
+			if self.height == top_frame.start_height {
+				// It is an error to pop more values than was pushed in the current frame
+				// (ie pop values pushed in the parent frame), unless the frame became
+				// polymorphic.
+				return if top_frame.is_polymorphic {
+					Ok(())
+				} else {
+					Err(Error("trying to pop more values than pushed".into()))
+				}
+			}
+		}
+
+		self.height = self.height
+			.checked_sub(value_count)
+			.ok_or_else(|| Error("stack underflow".into()))?;
+
+		Ok(())
+	}
+}
+
+/// This function expects the function to be validated.
+pub fn max_stack_height(func_idx: u32, module: &elements::Module) -> Result<u32, Error> {
```

Comment:
module called `stack_height`, submodule called `max_height`, function named `max_stack_height`, sure there is no better way to name all these?

Reply:
oh, this is binary, nevermind

Reply:
or make it pub(crate) ?